### DISCUSSION TITLE
Animated nav mesh? Or always static?

### DISCUSSION THREAD

**Discussion by jameskane05** [mozilla/hubs#2878] [General Questions & Discussion] [answered]
Is the nav mesh always static? Or can a piece of it be animated? We are thinking about a design that is more vertical, as if it's a multi-level high-rise. But we want players to be able to easily loop b/w the beginning and end so an animated looping elevator platform or ladder would be really nice (or even just a slow parachuting fall from the top). Has anyone seen or attempted anything like this? Definitely willing to hack toward it if anyone has pointers.

**Comment by netpro2k** [comment, marked as answer]
Hmm, could be tricky... I guess there is no reason you couldn't just mutate the navmesh data itself to make these moving platforms (we use https://github.com/donmccurdy/three-pathfinding for navmesh handling).. I think the character controller would mostly handle that correctly if you didn't move it too quickly, though I suspect you would see some jittering (not atypical in games on moving platforms).

I could also imagine 2 waypoint based solutions:
1. Clickable trigger to teleport to a waypoint. Basically make some sort of clickable object that when activated would teleport you to a waypoint. Should be pretty simple code wise, though its less "realistic".. Probably could come up with a cool way to present it visually though.
2. Use occupiable waypoints to "get in" an elevator. When you are "occupying" a waypoint, the normal navmesh snapping is turned off. If you put that waypoint on an animated node the user could be moved around with that waypoint (though this would require a very small code change, since right now the snapping only happens when you enter the waypoint instead of continuously). You would then need to teleport or walk off the elevator at the top (or maybe do some additional logic automatically move you off of it at a certain point).

The other issue you are going to run into with the moving platform approach is that animations are currently not networked, so without some additional work to sync them up, you would see people floating in the air if they were not synced up with your local representation of the elevator position.

**Comment by paradowskicreative** [comment]
I'm trying to prove out updating the nav mesh at runtime in the Hubs client. I've tried a few different approaches, but the most direct would seem to be calling the A-Frame system nav.loadMesh() and passing in a new Mesh (a simple plane for a test):

```
        let nav = document.querySelector('a-scene').systems["nav"];
        let geo = new THREE.PlaneBufferGeometry( 5, 20, 32 );
        var material = new THREE.MeshBasicMaterial( {color: 0xffff00, side: THREE.DoubleSide} );
        var plane = new THREE.Mesh( geo, material );
        nav.loadMesh(plane, "zone_name_1");
```

This runs w/o error and it has some effect - my character basically goes into camera/flying mode and teleportation stops working. So the existing nav mesh seems to be removed but there is no plane in its place. Do I have to reattach players to the new nav mesh after loading somehow? Any idea, @netpro2k ?
